MHArmory.Search/OpenCL/SearchDataSerializer.cs

```diff
@@ -0,0 +1,243 @@
+            {
+                throw new OpenCLSerializationException($"Tried to serialize {maps.JewelIdMap.Count} distinct jewels, maximum allowed is {SearchLimits.TotalJewels}");
+            }
+
+            maps.SetIdMap = CreateIDMap(desiredArmorSetParts.Select(x => x.Id));
+            if (maps.SetIdMap.Count > SearchLimits.MaxSetSkills)
+            {
+                throw new OpenCLSerializationException($"Tried to serialize {maps.SetIdMap.Count} distinct set parts, maximum allowed is {SearchLimits.MaxSetSkills}");
+            }
+
+            maps.SkillIdMap = CreateIDMap(desiredSkillIDs);
+            if (maps.SkillIdMap.Count > SearchLimits.MaxDesiredSkills)
+            {
+                throw new OpenCLSerializationException($"Tried to serialize {maps.SkillIdMap.Count} distinct skills, maximum allowed is {SearchLimits.MaxDesiredSkills}");
+            }
+
+            byte[] header = new byte[13];
+            SerializeWeaponSlots(data.WeaponSlots, header);
+            header[4] = (byte)heads.Count;
+            header[5] = (byte)chests.Count;
+            header[6] = (byte)gloves.Count;
+            header[7] = (byte)waists.Count;
+            header[8] = (byte)legs.Count;
+            header[9] = (byte)charms.Count;
+            header[10] = (byte)data.AllJewels.Length;
+            header[11] = (byte)data.DesiredAbilities.Length;
+            header[12] = (byte)maps.SetIdMap.Count;
+            parameters.Header = header;
+
+            parameters.Equipment = SerializeEquipment(maps, allEquipment);
+            parameters.DesiredSkills = SerializeDesiredSkills(maps, data.DesiredAbilities);
+            parameters.Decorations = SerializeDecorations(maps, data.AllJewels);
+            parameters.Combinations = (uint) (heads.Count * chests.Count * gloves.Count * waists.Count * legs.Count * charms.Count);
+            parameters.SearchIDMaps = maps;
+            return parameters;
+        }
+
+        private IDictionary<T, byte> CreateIDMap<T>(IEnumerable<T> keys)
+        {
+            byte currentIndex = 0;
+            var map = new Dictionary<T, byte>();
+            foreach (T key in keys)
+            {
+                if (!map.ContainsKey(key))
+                {
+                    map.Add(key, currentIndex);
+                    currentIndex++;
+                }
+            }
+            return map;
+        }
+
+        private void SerializeWeaponSlots(int[] weaponSlots, byte[] header)
+        {
+            foreach (int weaponSlot in weaponSlots)
+            {
+                header[weaponSlot]++;
+            }
+        }
+
+        private byte[] SerializeEquipment(SearchIDMaps maps, IEnumerable<IEquipment> equipments)
+        {
+            var ms = new MemoryStream();
+            var writer = new BinaryWriter(ms);
+            foreach (IEquipment equipment in equipments)
+            {
+                var tuple = new Tuple<int, EquipmentType>(equipment.Id, equipment.Type);
```

Comment:
Same as previously here, use the native tuple syntax, they are structures and thus are much more GC friendly.